packages/enzyme/src/ShallowWrapper.js

```diff
@@ -400,6 +402,10 @@ class ShallowWrapper {
           }
         }
         this.update();
+        // call the setState callback
+        if (callback) {
+          callback.call(instance);
```

Comment:
are we sure that React calls the callback with zero arguments, and the instance as the receiver?

Reply:
> are we sure that React calls the callback with zero arguments, and the instance as the receiver?

I think so.

* Implementation
    * https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberUpdateQueue.js#L352
* Demo
    * https://codepen.io/koba04/pen/XVpEEy?editors=0011

Reply:
That implementation doesn't show where `update.callback` is assigned; but your demo is convincing (for the latest version of React).

However, in React ^15.4, it seems to get one argument set to `undefined` in your demo (whereas in <= 15.3, and >= 16, it seems to get zero arguments). We should maintain that behavior via adapters.

Reply:
Thanks! I'll work on to support < 15.3

Reply:
I've tested for arguments of setState callback.

```js
setState({foo: 'bar'}, function(...args) {
  console.log(args);
});
```

* 0.13.3 ... `[]`
* 0.14.9 ... `[]`
* 15.4.2 ... `[ undefined ]`
* 15.6.2 ... `[ undefined ]`
* 16.2.0 ... `[]`

The arguments seem to be `[undefined]` only in v15.

Reply:
In my testing, it's `^15.4` - ie, >= 15.4 and < 16. If you check `15.3`, you should get `[]`.

Reply:
Thanks, you are right. I've updated the list.

* 0.13.3 ... `[]`
* 0.14.9 ... `[]`
* 15.3.2 ... `[]`
* 15.4.2 ... `[ undefined ]`
* 15.4.2 ... `[ undefined ]`
* 15.5.4 ... `[ undefined ]`
* 15.6.2 ... `[ undefined ]`
* 16.2.0 ... `[]`

Reply:
Changing the argument by depending on React's version seems to be painful.
So I'll try to update the implementation to use a spy method like  https://github.com/airbnb/enzyme/pull/1192 to know the argument.

Reply:
It's more than just painful; enzyme itself must not know about React at all, given that non-React adapters need to be able to work.

In other words, this functionality needs to be something that every adapter implements, and enzyme merely calls into.

Reply:
> enzyme itself must not know about React at all, given that non-React adapters need to be able to work.

I agree with you. But calling setState callback without argument is depending on current React implementation.
Spying setState callback and using the result later is to reproduce this functionality so I think it can be applied any adapters which are including non-React adapters.

```js
instance.setState(state, spy);
:
if (callback) {
  // replay setState callbacks
  spy.getCalls().forEach((spyCall) => {
    const context = spyCall.thisValue;
    const args = spyCall.args;
    callback.apply(context, args);
  });
}
```

Reply:
Your spy suggestion is interesting, but I was suggesting something like `adapter.invokeSetStateCallback(instance, callback)` - I'm not sure which is better.

Reply:
Thanks, I got it.
Are you ok to implement a default behavior into `EnzymeAdapter`? If not, it will be a breaking change.

Reply:
Yes, I think that would make sense.